api.js

```diff
@@ -720,6 +725,458 @@ authenticate.clientCache  = clientCache;
+            }
+          }
+
+          // Get credentials with ext
+          loadCredentials(clientId, ext, callback);
+        }, {}, authenticated);
+      }
+    });
+  };
+};
+
+/**
+ * Make a function for the remote signature validation.
+ *
+ * options:
+ * {
+ *    authBaseUrl:   'https://....' // baseUrl for authentication server
+ * }
+ *
+ * The function returns takes an object:
+ *     {method, url, host, port, authorization}
+ * And return promise for an object on one of the forms:
+ *     {message}, {scheme, scopes}, {schema, scopes, hash}
+ *
+ * The method returned by this function works as `signatureValidator` for
+ * `remoteAuthentication`.
+ */
+var createRemoteSignatureValidator = function(options) {
+  assert(options.authBaseUrl, "options.authBaseUrl is required");
+  var agent = null;
+  if (/^https:/.test(options.authBaseUrl)) {
+    agent = new https.Agent({
+      keepAlive:  true,
+      maxSockets: 50
+    });
+  } else {
+    agent = new http.Agent({
+      keepAlive:  true,
+      maxSockets: 50
+    });
+  }
+  return function(data) {
+    return request
+      .post(options.authBaseUrl + '/authenticate-hawk')
+      .agent(agent)
+      .type('json')
+      //TODO: Use TC-client when we have auth.tc.net deployed...
+      //      So we get both retries and decent agent implementation
+      .send(data)
+      .end()
+      .then(function(res) {
+        if (!res.ok) {
+          throw new Error(
+            "Received status code: " + res.status + " from /authenticate-hawk"
+          );
+        }
+        return res.body;
+      });
+  };
+};
+
+/**
+ * Validate signature using remote API end-point
+ *
+ * options:
+ * {
+ *    signatureValidator:   async (data) => {message}, {scheme, scopes}, or
+ *                                          {schema, scopes, hash}
+ * }
+ *
+ * where `data` is the form: {method, url, host, port, authorization}.
+ */
+var remoteAuthentication = function(options, entry) {
+  assert(options.signatureValidator instanceof Function,
+         "Expected options.signatureValidator to be a function!");
+  // Returns promise for object on the form: {error, message, scopes}
+  var authenticate = function(req) {
+    // Check that we're not using two authentication schemes, we could
+    // technically allow two. There are cases where we redirect and it would be
+    // smart to let bewit overwrite header authentication.
+    // But neither Azure or AWS tolerates two authentication schemes,
+    // so this is probably a fair policy for now. We can always allow more.
+    if (req.headers && req.headers.authorization &&
+        req.query && req.query.bewit) {
+      return Promise.resolve({
+        message:  "Cannot use two authentication schemes at once " +
+                  "this request has both bewit in querystring and " +
+                  "and 'authorization' header"
+      });
+    }
+
+    // If no authentication is provided, we just return valid with zero scopes
```

Comment:
@djmitche It looked alright to me, since it's going to return {status: 'no-auth', ... , scopes: []} what am I missing?